core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java

```diff
@@ -502,6 +502,11 @@ private RelDataType getTightestCommonTypeOrThrow(
       return null;
     }
 
+    if (type1 instanceof RelDataTypeFactoryImpl.JavaType
```

Comment:
this means that no casts will be inserted. is this the desired behavior?

Reply:
@mihaibudiu 
Hi, I found this problem in the ReflectiveSchemaTest#testCompareJavaAndSqlTypes.
Now in the condition `primitiveLong > 0`, the primitiveLong will be cast to RelDataType.INTEGER. This will cause value overflow(Now don't, because we don't add CAST in primitiveLong).
According to AbstractTypeCoercion,#needToCast prevents casting a JavaType to a normal RelDataType, so we don't need support to convert it.
Before CALCITE-6617, the ReflectiveSchemaTest#testCompareJavaAndSqlTypes returned `null` when comparing a Java type with an SQL type INTEGER.
Now the ReflectiveSchemaTest#testCompareJavaAndSqlTypes can't handle it when the input is Java type and Java type. Because no exception was thrown in the unit test, So I don't support it now.
﻿

Reply:
This discussion is very useful, can you please add a little bit of this as a comment to this method, so other people who may wonder later have an answer ready?

Reply:
Done!